R/data.table.R

```diff
@@ -1324,8 +1324,21 @@ replace_dot_alias = function(e) {
           ## check key on i as well!
           ichk = is.data.table(i) && haskey(i) &&
                  identical(head(key(i), length(leftcols)), names_i[leftcols]) # i has the correct key, #3061
-          if (keylen && (ichk || is.logical(i) || (.Call(CisOrderedSubset, irows, nrow(x)) && ((roll == FALSE) || length(irows) == 1L)))) # see #1010. don't set key when i has no key, but irows is ordered and roll != FALSE
-            setattr(ans,"sorted",head(key(x),keylen))
+          if (
+            keylen
+            && (
+              ichk
+              || is.logical(i)
+              || (
+                .Call(CisOrderedSubset, irows, nrow(x))
+                && (!roll || length(irows) == 1L) # see #1010. don't set key when i has no key, but irows is ordered and roll != FALSE
+                && (  #5361 merging on keyed factor with character, check if resulting character is really sorted
+                  identical(vapply_1c(i[,leftcols,with=FALSE], typeof), vapply_1c(x[,rightcols,with=FALSE], typeof)) # can only be not identical
```

Comment:
Shouldn't we be able to check the types of columns in `i` and `x` without making a copy (some columns) of the two tables? e.g. `vapply_1c(left_cols, function(col) typeof(i[[col]]))`?

Reply:
Good point. It seems also to be faster!

```r
set.seed(1L)
dt = as.data.table(matrix(rnorm(1e8), nrow=1e4))
cols = sample(1e4)
bench::mark(
    unname(vapply_1c(dt[, cols, with=FALSE], typeof)), 
    vapply_1c(cols, function(col) typeof(dt[[col]]))
)
#> # A tibble: 2 × 13
#>   expression     min  median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
#>   <bch:expr>  <bch:> <bch:t>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
#> 1 unname(vap… 93.2ms 133.7ms      7.51   764.1MB     9.38     4     5      533ms
#> 2 vapply_1c(… 28.2ms  28.9ms     30.5     84.8KB    38.1     16    20      525ms
#> # ℹ 4 more variables: result <list>, memory <list>, time <list>, gc <list>
```

Reply:
did not expect it to be an order of magnitude faster!